src/v/config/constraint_store.h

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "config/configuration.h"
+#include "config/constraint_types.h"
+
+#include <fmt/format.h>
+
+#include <unordered_map>
+
+namespace config {
+// Similar to config_store
+class constraint_store {
+public:
+    ~constraint_store() = default;
+
+    bool contains(ss::sstring& name) { return _constraints.contains(name); }
+
+    constraint& get(const ss::sstring& name) {
+        auto found = _constraints.find(name);
+        if (found != _constraints.end()) {
+            return found->second;
+        } else {
+            throw std::out_of_range(
+              fmt::format("Constraint {} not found", name));
+        }
+    }
+
+    /**
+     * Creates constraints based on the constraint mapping from the
+     * configuration store. This implies that the broker already read the YAML
+     * config file.
+     */
+    void load_constraints() {
+        load_constraints(
+          config::shard_local_cfg().constraints(), config::shard_local_cfg());
+    }
+
+    void load_constraints(
+      const std::unordered_map<constraint_args::key_type, constraint_args>&
+        constraint_args,
+      config::config_store& conf) {
+        _constraints.reserve(constraint_args.size());
+        for (const auto& [name, args] : constraint_args) {
+            auto& property = conf.get(name);
+            constraint cons{
+              std::move(args), property.make_constraint_methods()};
+            _constraints[name] = std::move(cons);
```

Comment:
The move doesn't do anything, and `insert_or_assign` avoids the need for a default constructor.
```suggestion
            constraint cons{args, property.make_constraint_methods()};
            _constraints.insert_or_assign(name, std::move(cons));
```